components/vc/front/src/torque.c

```diff
@@ -47,13 +54,167 @@ static struct
+    {
+        torque_data.gear = torque_data.gear == GEAR_F ? GEAR_R : GEAR_F;
+    }
+#endif
+}
+
+static void evaluate_mode_change(void)
+{
+    CAN_digitalStatus_E race_mode_change_requested = CAN_DIGITALSTATUS_SNA;
+    const bool race_mode_change_was_requested = torque_data.race_mode_change_active;
+    torque_data.race_mode_change_active = (CANRX_get_signal(VEH, SWS_requestRaceMode, &race_mode_change_requested) != CANRX_MESSAGE_SNA) &&
+                                          (race_mode_change_requested == CAN_DIGITALSTATUS_ON);
+    const bool race_mode_change_rising = !race_mode_change_was_requested && torque_data.race_mode_change_active;
+
+    if (race_mode_change_rising)
+    {
+        torque_data.race_mode = torque_data.race_mode == RACEMODE_ENABLED ? RACEMODE_PIT : RACEMODE_ENABLED;
+    }
+}
+
+static float32_t evaluate_torque_max(void)
+{
+    float32_t torque_request_max = torque_data.torque_request_max;
+    CAN_digitalStatus_E torque_change_request = CAN_DIGITALSTATUS_SNA;
+    const bool torque_inc_active = (CANRX_get_signal(VEH, SWS_requestTorqueInc, &torque_change_request) != CANRX_MESSAGE_SNA) &&
+                                   (torque_change_request == CAN_DIGITALSTATUS_ON);
+    const bool torque_dec_active = (CANRX_get_signal(VEH, SWS_requestTorqueDec, &torque_change_request) != CANRX_MESSAGE_SNA) &&
+                                   (torque_change_request == CAN_DIGITALSTATUS_ON);
+
+    if (torque_inc_active ^ torque_dec_active)
+    {
+        const drv_timer_state_E timer_state = drv_timer_getState(&torque_data.torque_change_timer);
+        if (timer_state == DRV_TIMER_STOPPED)
+        {
+            drv_timer_start(&torque_data.torque_change_timer, TORQUE_CHANGE_DELAY);
+            torque_request_max = torque_inc_active ? torque_request_max + 1 : torque_request_max - 1;
+            torque_request_max = SATURATE(MIN_TORQUE_RANGE, torque_request_max, ABSOLUTE_MAX_TORQUE);
+        }
+        else if (timer_state == DRV_TIMER_EXPIRED)
+        {
+            drv_timer_stop(&torque_data.torque_change_timer);
+        }
+
+        torque_data.torque_request_max = torque_request_max;
+    }
+    else
+    {
+        drv_timer_stop(&torque_data.torque_change_timer);
+    }
+
+    return torque_request_max;
+}
+
+static void evaluate_launch_control(float32_t accelerator_position, float32_t brake_position)
+{
+#if FEATURE_IS_ENABLED(FEATURE_LAUNCH_CONTROL)
+    CAN_digitalStatus_E launch_control_requested = CAN_DIGITALSTATUS_SNA;
+    const bool launch_control_request_was_active = torque_data.torque_control_request_active;
+    torque_data.torque_control_request_active = (CANRX_get_signal(VEH, SWS_requestLaunchControl, &launch_control_requested) != CANRX_MESSAGE_SNA) &&
+                                                (launch_control_requested == CAN_DIGITALSTATUS_ON);
+    const bool launch_control_request_rising = !launch_control_request_was_active && torque_data.torque_control_request_active;
```

Comment:
do you want to do this check all the time? Or just if you're in the INACTIVE state?